library/ssl_tls13_client.c

```diff
@@ -93,35 +94,375 @@ static int ssl_tls13_write_supported_versions_ext( mbedtls_ssl_context *ssl,
+static int ssl_tls13_write_named_group_dhe( mbedtls_ssl_context *ssl,
+                                              unsigned char *buf,
+                                              unsigned char *end,
+                                              size_t *olen )
 {
     ((void) ssl);
     ((void) buf);
     ((void) end);
-    ((void) olen);
+    *olen = 0;
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "write_named_group_dhe is not implemented" ) );
     return( MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE );
 }
 
-#endif /* MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED */
+/*
+ * Supported Groups Extension (supported_groups)
+ */
+static int ssl_tls13_write_supported_groups_ext( mbedtls_ssl_context *ssl,
+                                                 unsigned char *buf,
+                                                 unsigned char *end,
+                                                 size_t *olen )
+{
+    unsigned char *p = buf ;
+    unsigned char *named_group_ptr; /* Start of named_group_list */
+    size_t named_group_len = 0;
+    int ret = 0, ret_ecdhe, ret_dhe;
+
+    *olen = 0;
+
+    if( !mbedtls_ssl_conf_tls13_some_ephemeral_enabled( ssl ) )
+        return( 0 );
+
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello, adding supported_groups extension" ) );
+
+    /* Check there is space for extension header */
+    MBEDTLS_SSL_CHK_BUF_PTR( p, end, 6 );
+    p += 6;
+
+    named_group_ptr = p;
+    ret_ecdhe = ssl_tls13_write_named_group_ecdhe( ssl, p, end, &named_group_len );
+    if( ret_ecdhe != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_RET( 1, "ssl_tls13_write_named_group_ecdhe", ret );
+    }
+    p += named_group_len;
+
+    ret_dhe = ssl_tls13_write_named_group_dhe( ssl, p, end, &named_group_len );
+    if( ret_dhe != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_RET( 1, "ssl_tls13_write_named_group_dhe", ret );
+    }
+    p += named_group_len;
+
+    /* Both ECDHE and DHE Fail. */
+    if( ret_ecdhe != 0 && ret_dhe != 0 )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "Both ECDHE and DHE groups are fail. " ) );
+        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+    }
+
+    /* Length of named_group_list*/
+    named_group_len = p - named_group_ptr;
+    if( named_group_len == 0 )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "No Named Group Available." ) );
+        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+    }
+
+    /* Write extension_type */
+    MBEDTLS_PUT_UINT16_BE( MBEDTLS_TLS_EXT_SUPPORTED_GROUPS, buf, 0 );
+    /* Write extension_data_length */
+    MBEDTLS_PUT_UINT16_BE( named_group_len + 2, buf, 2 );
+    /* Write length of named_group_list */
+    MBEDTLS_PUT_UINT16_BE( named_group_len, buf, 4 );
+
+    MBEDTLS_SSL_DEBUG_BUF( 3, "Supported groups extension", buf + 4, named_group_len + 2 );
+
+    *olen = p - buf;
+
+    ssl->handshake->extensions_present |= MBEDTLS_SSL_EXT_SUPPORTED_GROUPS;
+
+    return( ret );
+}
 
 /*
- * Functions for writing ClientHello message.
+ * Functions for writing key_share extension.
  */
+#if defined(MBEDTLS_ECDH_C)
```

Comment:
In `prototype` , the macro condition is `#if defined(MBEDTLS_ECDH_C) || defined(MBEDTLS_ECDSA_C)` .
But it can not pass CI test under ` !MBEDTLS_ECDH_C  && MBEDTLS_ECDSA_C ` .  So I remove `MBEDTLS_ECDSA_C` . 
I am not sure if it is correct, please confirm that